File: App/app/src/main/java/com/cinema/ticket_booking/ui/notification/NotificationFragment.java
package com.cinema.ticket_booking.ui.notification;

import android.os.Bundle;
import com.cinema.ticket_booking.util.SnackbarHelper;
import com.cinema.ticket_booking.databinding.FragmentNotificationBinding;
import com.cinema.ticket_booking.ui.notification.NotificationAdapter;
import com.google.android.material.snackbar.Snackbar;
import android.view.*;
import android.widget.Toast;
import androidx.annotation.*;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.*;
import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class NotificationFragment extends Fragment {

    private FragmentNotificationBinding binding;
    private NotificationViewModel viewModel;

    @Override
    public View onCreateView(@NonNull LayoutInflater i, ViewGroup c, Bundle s) {
        binding = FragmentNotificationBinding.inflate(i, c, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(this).get(NotificationViewModel.class);

        binding.rvNotifications.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.btnBack.setOnClickListener(v -> Navigation.findNavController(view).popBackStack());
        binding.btnMarkAllRead.setOnClickListener(v -> {
            viewModel.markAllAsRead();
            SnackbarHelper.showSuccess(binding.getRoot(), "Đã đánh dấu tất cả là đã đọc");
        });

        viewModel.getNotifications().observe(getViewLifecycleOwner(), resource -> {
            switch (resource.status) {
                case LOADING -> binding.progressBar.setVisibility(View.VISIBLE);
                case SUCCESS -> {
                    binding.progressBar.setVisibility(View.GONE);
                    if (resource.data == null || resource.data.getContent() == null
                            || resource.data.getContent().isEmpty()) {
                        binding.tvEmpty.setVisibility(View.VISIBLE);
                    } else {
                        binding.tvEmpty.setVisibility(View.GONE);
                        NotificationAdapter adapter = new NotificationAdapter(resource.data.getContent());
                        binding.rvNotifications.setAdapter(adapter);
                        setupSwipeToDelete(adapter);
                    }
                }
                case ERROR -> {
                    binding.progressBar.setVisibility(View.GONE);
                    SnackbarHelper.showError(binding.getRoot(), resource.message);
                }
            }
        });
    }

    private void setupSwipeToDelete(NotificationAdapter adapter) {
        new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(@NonNull RecyclerView rv, @NonNull RecyclerView.ViewHolder vh,
                    @NonNull RecyclerView.ViewHolder t) {
                return false;
            }

            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder vh, int dir) {
                int pos = vh.getAdapterPosition();
                var item = adapter.getItem(pos);

                // Tạm thời xóa khỏi UI
                adapter.removeItem(pos);

                // Hiện Snackbar với nút Undo
                Snackbar snackbar = Snackbar.make(
                        binding.getRoot(), "Đã xóa thông báo", Snackbar.LENGTH_LONG);

                boolean[] undoPressed = { false };
                snackbar.setAction("Hoàn tác", v -> {
                    undoPressed[0] = true;
                    // Tải lại dữ liệu (đơn giản nhất) hoặc insert lại adapter
                    viewModel.getNotifications().observe(getViewLifecycleOwner(), r -> {
                        if (r.status == com.cinema.ticket_booking.util.Resource.Status.SUCCESS) {
                            binding.rvNotifications.setAdapter(new NotificationAdapter(r.data.getContent()));
                        }
                    });
                });

                snackbar.addCallback(new Snackbar.Callback() {
                    @Override
                    public void onDismissed(Snackbar transientBottomBar, int event) {
                        if (!undoPressed[0]) {
                            // Nếu không Undo thì mới gọi API xóa thật
                            viewModel.deleteNotification(item.getId().toString());
                        }
                    }
                });

                snackbar.show();
            }
        }).attachToRecyclerView(binding.rvNotifications);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/notification/NotificationFragment.kt
package com.cinema.ticket_booking.ui.notification

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.cinema.ticket_booking.databinding.FragmentNotificationBinding
import com.cinema.ticket_booking.util.Resource
import com.cinema.ticket_booking.util.SnackbarHelper
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class NotificationFragment : Fragment() {

    private var _binding: FragmentNotificationBinding? = null
    private val binding get() = _binding!!
    private lateinit var viewModel: NotificationViewModel

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentNotificationBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this)[NotificationViewModel::class.java]

        binding.rvNotifications.layoutManager = LinearLayoutManager(requireContext())
        binding.btnBack.setOnClickListener { v -> Navigation.findNavController(view).popBackStack() }
        binding.btnMarkAllRead.setOnClickListener {
            viewModel.markAllAsRead()
            SnackbarHelper.showSuccess(binding.root, "Đã đánh dấu tất cả là đã đọc")
        }

        viewModel.notifications.observe(viewLifecycleOwner) { resource ->
            when (resource.status) {
                Resource.Status.LOADING -> binding.progressBar.visibility = View.VISIBLE
                Resource.Status.SUCCESS -> {
                    binding.progressBar.visibility = View.GONE
                    val content = resource.data?.content
                    if (content.isNullOrEmpty()) {
                        binding.tvEmpty.visibility = View.VISIBLE
                    } else {
                        binding.tvEmpty.visibility = View.GONE
                        val adapter = NotificationAdapter(content.toMutableList())
                        binding.rvNotifications.adapter = adapter
                        setupSwipeToDelete(adapter)
                    }
                }
                Resource.Status.ERROR -> {
                    binding.progressBar.visibility = View.GONE
                    SnackbarHelper.showError(binding.root, resource.message ?: "Có lỗi xảy ra")
                }
            }
        }
    }

    private fun setupSwipeToDelete(adapter: NotificationAdapter) {
        ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
            override fun onMove(
                rv: RecyclerView,
                vh: RecyclerView.ViewHolder,
                t: RecyclerView.ViewHolder
            ): Boolean {
                return false
            }

            override fun onSwiped(vh: RecyclerView.ViewHolder, dir: Int) {
                val pos = vh.adapterPosition
                val item = adapter.getItem(pos)

                // Tạm thời xóa khỏi UI
                adapter.removeItem(pos)

                // Hiện Snackbar với nút Undo
                val snackbar = Snackbar.make(binding.root, "Đã xóa thông báo", Snackbar.LENGTH_LONG)
                val undoPressed = booleanArrayOf(false)

                snackbar.setAction("Hoàn tác") {
                    undoPressed[0] = true
                    // Tải lại dữ liệu (đơn giản nhất) hoặc insert lại adapter
                    viewModel.notifications.observe(viewLifecycleOwner) { r ->
                        if (r.status == Resource.Status.SUCCESS) {
                            val content = r.data?.content
                            val list = content?.toMutableList() ?: mutableListOf()
                            binding.rvNotifications.adapter = NotificationAdapter(list)
                        }
                    }
                }

                snackbar.addCallback(object : Snackbar.Callback() {
                    override fun onDismissed(transientBottomBar: Snackbar, event: Int) {
                        if (!undoPressed[0]) {
                            // Nếu không Undo thì mới gọi API xóa thật
                            item.id?.toString()?.let { viewModel.deleteNotification(it) }
                        }
                    }
                })

                snackbar.show()
            }
        }).attachToRecyclerView(binding.rvNotifications)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
